pkg/sql/ttl/ttljob/ttljob_test.go

```diff
@@ -8,33 +8,348 @@
+				return th.cfg.DB.Txn(context.Background(), func(ctx context.Context, txn *kv.Txn) error {
+					return fn(ctx, 0 /* allSchedules */, txn)
+				})
+			}
+		},
+
+		CaptureJobExecutionConfig: func(config *scheduledjobs.JobExecutionConfig) {
+			th.cfg = config
+		},
+	}
+
+	var zeroDuration time.Duration
+	args := base.TestServerArgs{
+		Knobs: base.TestingKnobs{
+			JobsTestingKnobs: knobs,
+			TTL: &sql.TTLTestingKnobs{
+				AOSTDuration: &zeroDuration,
+			},
+		},
+	}
+	s, db, kvDB := serverutils.StartServer(t, args)
+	require.NotNil(t, th.cfg)
+	th.kvDB = kvDB
+	th.sqlDB = sqlutils.MakeSQLRunner(db)
+	th.server = s
+
+	return th, func() {
+		s.Stopper().Stop(context.Background())
+	}
+}
+
+func (h *rowLevelTTLTestJobTestHelper) waitForSuccessfulScheduledJob(t *testing.T) {
+	query := fmt.Sprintf(
+		`SELECT status, error FROM [SHOW JOBS] 
+		WHERE job_id IN (
+			SELECT job_id FROM %s
+			WHERE created_by_id IN (SELECT schedule_id FROM %s WHERE executor_type = 'scheduled-row-level-ttl-executor')
+		)`,
+		h.env.SystemJobsTableName(),
+		h.env.ScheduledJobsTableName(),
+	)
+
+	testutils.SucceedsSoon(t, func() error {
+		// Force newly created job to be adopted and verify it succeeds.
+		h.server.JobRegistry().(*jobs.Registry).TestingNudgeAdoptionQueue()
+		var status, errorStr string
+		if err := h.sqlDB.DB.QueryRowContext(
+			context.Background(),
+			query,
+		).Scan(&status, &errorStr); err != nil {
+			return errors.Wrapf(err, "expected to scan row for a job, got")
+		}
+
+		if status != string(jobs.StatusSucceeded) {
+			return errors.Newf("expected status %s, got %s (error: %s)", jobs.StatusSucceeded, status, errorStr)
+		}
+		return nil
+	})
+}
+
+// TestRowLevelTTLJobRandomEntries inserts random entries into a given table
+// and runs a TTL job on them.
+func TestRowLevelTTLJobRandomEntries(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	testCases := []struct {
-		cols     []string
-		expected string
-	}{
-		{[]string{"a"}, "a"},
-		{[]string{"index"}, `"index"`},
-		{[]string{"a", "b"}, "a, b"},
-		{[]string{"escape-me", "index", "c"}, `"escape-me", "index", c`},
+	rng, _ := randutil.NewTestRand()
+
+	var indexableTyps []*types.T
+	for _, typ := range types.Scalar {
+		// TODO(#76419): DateFamily has a broken `-infinity` case.
+		if colinfo.ColumnTypeIsIndexable(typ) && typ.Family() != types.DateFamily {
+			indexableTyps = append(indexableTyps, typ)
+		}
+	}
+
+	type testCase struct {
+		desc              string
+		createTable       string
+		numExpiredRows    int
+		numNonExpiredRows int
+		numSplits         int
+	}
+	// Add some basic one and three column row-level TTL tests.
+	testCases := []testCase{
+		{
+			desc: "one column pk",
+			createTable: `CREATE TABLE tbl (
+	id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+	text TEXT
+) WITH (ttl_expire_after = '30 days')`,
+			numExpiredRows:    1001,
+			numNonExpiredRows: 5,
+		},
+		{
+			desc: "one column pk, concurrent",
```

Comment:
this file was a move from `ttljob_integration_test.go`; only the concurrent tests are new